Configurable readiness probe for composed resources

muvaf opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

This PR introduces `readinessProbe` field to base templates in `Composition` resource so that users can configure how they would like that specific composed resource should be considered ready or not.

Fixes #1687 

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Manually.

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

muvaf commented:
Keeping it as draft until unit tests are written.

negz commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -87,6 +87,26 @@ type ComposedTemplate struct {
+
+	// ReadinessProbe allows users to use a custom readiness check. The default
+	// readiness probe is to have the "Ready" condition to be "True".
+	ReadinessProbe *ReadinessProbe `json:"readinessProbe,omitempty"`
```
Naming nit: "probe" to me implies that we'd be "probing" the resource whose readiness we're concerned with - i.e. doing something a little more active like running an exec or making a network request. I'd prefer something like `readinessConstraint`, `readinessRequirement`, or `readinessCheck` here.

negz replied:
Also, I think we should support an array of these. I'm very skeptical that inspecting only one field to determine that something is ready will scale over time.

negz commented on apis/apiextensions/v1alpha1/composition_types.go:
```diff
@@ -87,6 +87,26 @@ type ComposedTemplate struct {
+
+	// Type indicates the type of probe you'd like to use.
+	// +kubebuilder:validation:Enum="Match";"NonEmpty"
+	Type string `json:"type"`
```
Can we make a subtype of `string` - i.e. `ReadinessProbeType` - with the possible values as constants in this file?

negz commented on pkg/controller/apiextensions/composite/composed/api.go:
```diff
@@ -164,3 +167,36 @@ func (cdf *APIConnectionDetailsFetcher) Fetch(ctx context.Context, cd resource.C
+
+// DefaultReadinessProber is a readiness prober which returns whether the composed
+// resource is ready or not.
+type DefaultReadinessProber struct{}
```
Nit: Could we use a `ReadinessProberFn`, rather than using an empty struct named `Default<T>` to satisfy our interface? It doesn't seem like this needs to be a struct.

Edit: Ah, I see this is an established convention in this file. It's not my preferred convention, but I'm happy to stick to it.

muvaf replied:
That would've been nicer indeed. I want to defer it to later though.

muvaf commented:
@negz addressed all of your comments. It's ready to merge but we need https://github.com/crossplane/crossplane-runtime/pull/199 merged and update `go.mod` of this file. I've manually tested all the functionality with the composition guide we have at crossplane.io

negz commented on go.mod:
```diff
@@ -2,6 +2,8 @@ module github.com/crossplane/crossplane
 
 go 1.13
 
+replace github.com/crossplane/crossplane-runtime => github.com/muvaf/crossplane-runtime v0.0.0-20200910111027-446c12e8ee41
```
Please remove and update to master before merging.

negz reviewed (approved):
Thanks @muvaf this looks great. Feel free to merge once the `replace` statement has been removed.
